Add user docs for molecular and periodic DF (#17)
pyscf/pyscf.github.io: source/user/df.rst

*Modules*: :mod:`df`, :mod:`pbc.df`

.. module:: df
   :synopsis: Density fitting and RI approximation
.. sectionauthor:: Qiming Sun <osirpt.sun@gmail.com>.

Introduction
============

> Density fitting (DF), sometimes also called the resolution of identity (RI) approximation, is a method to approximate the four-index electron repulsion integrals (ERIs) by two- and three-index tensors. :cite:`Dunlap00PCCP` It achieves so by expanding the atomic orbital (AO) pair density in a second, auxiliary Gaussian basis set. PySCF has extensive supports for using DF in both molecular and crystalline calculations. This section of the user document covers the basic usage of DF in molecular applications. See :ref:`theory_pbc_df` for the use of DF in periodic calculations.

Reviewer: The sentence "It achieves so by expanding the atomic orbital (AO) pair density in a second, auxiliary Gaussian basis set." is weird.

I would replace it with something like:
"In density fitting, the degenerate atomic orbital (AO) product space is expanded in terms of an auxiliary basis set."

Author: By "degenerate" you mean sth like "redundant" or "over-complete"?

Reviewer: Yes, the set of basis function products is heavily linearly dependent.

Author: fixed

Using DF in calculations
========================

DF is not used by default when initializing a :class:`SCF` or :class:`MCSCF` object, but can be invoked via the :func:`density_fit` method as follows::

    from pyscf import gto, scf, mcscf
    mol = gto.M(atom='N 0 0 0; N 0 0 1.2', basis='def2-tzvp')
    mf = scf.RHF(mol).density_fit().run()
    mc = mcscf.CASSCF(mf, 8, 10).density_fit().run()

Alternatively, one can initialize a :class:`DF` object and overwrite the :attr:`with_df` attribute of the :class:`SCF` object::

    from pyscf import gto, scf, df
    mol = gto.M(atom='N 0 0 0; N 0 0 1.2', basis='def2-tzvp')
    mydf = df.DF(mol).build()
    mf = scf.RHF(mol).density_fit()
    mf.with_df = mydf
    mf.kernel()

Setting :attr:`with_df` to ``None`` turns DF off.

> Once the SCF calculations using DF are done, all post-SCF methods will use DF automatically. The example below performs a TDDFT/TDA calculation using DF::

Reviewer: This is also weird: "once the SCF calculations using DF are done".

I think the point is that if you run a post-SCF calculation on top of a DF calculation, the post-SCF calculation will also use density fitting.

Reviewer: This also leads to a question of accuracy. Let's say I run an RI-HF calculation, and then proceed with CCSD. What's the auxiliary basis used for the CCSD part? An RI-JK basis will be accurate for RI-HF, but I'm not sure if it will automatically be okay for post-HF...

Author: Good question. I think at least in our group when doing PBC calculations, people tend to use the same basis for mean-field and post-HF calculations (i.e., JKFIT if cc-pVXZ is used). But I could be wrong and @tberkel can correct me.
I'm wondering how big a *practical* difference it will make by using e.g., a JKFIT basis for MP2 or CCSD compared to using the RI basis. @susilehtola do you know of any classical papers on this topic?

Reviewer: I'm not 100% sure; there are RIMP2 auxiliary basis sets that are specifically aimed for post-HF calculations for e.g. both the Karlsruhe and the Dunning families.

Reviewer: Weigend never disappoints
http://doi.org/10.1063/1.1445115

Reviewer: Here's the same for Karlsruhe https://doi.org/10.1039/C4CP04286G

Author: Thanks! I added an example using the `cc-pVTZ-RI` basis for DF-MP2 on top of a HF calculation that does *not* use DF.

    from pyscf import gto, dft, tddft
    mol = gto.M(atom='N 0 0 0; N 0 0 1.2', basis='def2-tzvp')
    mf = dft.RKS(mol, xc="b3lyp").density_fit().run()
    td = tddft.TDA(mf).run()

>     print(td.e)

Reviewer: What's the auxiliary basis that ends up used here?

Author: Good question! I checked the source code of the TDDFT module and didn't find any trace of DF being used explicitly. For example, when ERI is needed as in [this line](https://github.com/pyscf/pyscf/blob/master/pyscf/tdscf/rhf.py#L137), it directly calls the `ao2mo`.

Author: I just DM-ed qiming on this and will make revisions accordingly.

.. _choice_of_auxbasis:

Choice of auxiliary basis
=========================

General consideration
---------------------

The choice of auxiliary basis depends on the AO basis.

> By default, PySCF uses the pre-defined auxiliary basis set optimized for the AO basis set if the former exists.

Reviewer: ```suggestion
By default, PySCF uses a pre-defined auxiliary basis set optimized for the used AO basis set, if one exists.
```

Author: fixed

This includes many commonly used AO basis sets in electronic structure calculations, e.g., the Ahlrichs' `def2` family, :cite:`Hellweg07TCA,Weigend98CPL` the Dunning's `cc` family, :cite:`Weigend02PCCP` etc.

> When pre-defined auxiliary basis set is not available, an even-tempered basis (ETB) set is generated by the following rule

Reviewer: when a

Author: fixed

.. math::

    \varphi &= r^l \exp(-\zeta_{il} r^2), \quad i = 0, 1, \cdots, n \\
    \zeta_{il} &= \alpha_l \times \beta^i

> where :math:`\alpha_l` and :math:`N` are inferred automatically from the AO basis, and :math:`\beta = 2.0` by default.

Reviewer: how?

Author: Added explanation.

The user can overwrite the default choice by setting :attr:`auxbasis` upon initialization or at a later stage::

    from pyscf import gto, scf, df
    mol = gto.M(atom='N 0 0 0; N 0 0 1.2', basis='cc-pvdz')
    mf = scf.RHF(mol).density_fit(auxbasis="weigend")
    mf.kernel() # -108.910953335055
    mf.with_df.auxbasis = "cc-pvdz-jkfit"   # this is the default
    mf.kernel() # -108.913710743723
    mf.with_df.auxbasis = df.aug_etb(mol, beta=1.7) # ETB with beta = 1.7
    mf.kernel() # -108.914059329528

> The user can check what auxiliary basis set is used at any time by::

Reviewer: ```suggestion
The user can print the used auxiliary basis set at any time by::
```

Author: fixed

    print(mf.with_df.auxmol.basis)

More examples on using customized auxiliary basis can be found in :source:`examples/df/01-auxbasis.py`.

Special consideration for DFT
-----------------------------

> For DFT calculations with pure exchange-correlation functionals (i.e., LDA and GGA), the default auxiliary basis, which is designed for fitting both the Coulomb and the Hartree-Fock exchange integrals, may be unnecessarily too large. We recommend using the ``weigend`` family :cite:`Weigend06PCCP` for a more cost-effective choice as shown in the following example::

Reviewer: unnecessarily large

Author: fixed

    from pyscf import gto, dft
    mol = gto.M(atom='N 0 0 0; N 0 0 1.2', basis='def2-tzvpp')
    mf = dft.RKS(mol, xc="pbe").density_fit().run() # -109.432329411505
    print(mf.with_df.auxmol.basis)                  # default: def2-tzvpp-jkfit
    print(mf.with_df.auxmol.nao_nr())               # 154 aux basis functions

>     mf.with_df.auxbasis = "weigend"

Reviewer: "weigend" should be `def2-universal-jfit`, defined in Weigend2006

Reviewer: and one can note that the alternative for calculations including exact exchange is `def2-universal-jkfit`, defined in Weigend2008 (doi: 10.1002/jcc.20702)

Author: Changed to `def2-universal-jfit` and added a note that this is same as using `weigend`.
For the `def2-universal-jkfit`, I think that is the default used by PySCF if AO basis from the `def2` family is used.

    mf.kernel()                                     # -109.432334646585
    print(mf.with_df.auxmol.nao_nr())               # 98 aux basis functions

.. _save_reuse_df_integrals:

Saving and reusing DF integrals
===============================

> By default, the DF integrals are discarded after the calculations are done. Sometimes it is beneficial to save those integrals on disk and re-use them for later calculations. This can be done by specifying a HDF5 file where the DF integrals will be saved via setting :attr:`_cderi_to_save` either at the SCF stage::

Reviewer: ```suggestion
By default, the DF integrals are discarded once the calculation finishes.
Sometimes it is useful to save the integrals on disk for re-use in later calculations.
This can be achieved by specifying a HDF5 file to save the DF integrals in by setting :attr:`_cderi_to_save` either at the SCF stage::
```

Reviewer: Are these the DF two- and three-center integrals, or the B-matrices?

Author: it's the 3-index cholesky decomposed electron repulsion integrals, CDERIs as it's called in PySCF (which is your B-matrices I believe).
added text explaining this.

Reviewer: Yes; B-matrices factorize the integrals `(uv|rs) = sum_P B_{uv}^P B_{rs}^P`, but density fitted B matrices are still not the same as true Cholesky since the former have a fitting error while the latter can be made exact.

It would be cool to have the exact Cholesky decomposition in PySCF...

    mf = scf.RHF(mol).density_fit()
    mf.with_df._cderi_to_save = 'saved_cderi.h5'
    mf.kernel()

or initializing a :class:`DF` object separately::

    mydf = df.DF(mol)
    mydf.auxbasis = df.make_auxbasis(mol)
    mydf._cderi_to_save = 'saved_cderi.h5'
    mydf.build()

> The save integrals can be used later by setting :attr:`_cderi` to the HDF5 file holding the DF integrals::

Reviewer: saved

Author: fixed
